packages/flutter/lib/src/painting/box_border.dart

```diff
@@ -603,50 +637,59 @@ class Border extends BoxBorder {
       }
     }
 
-    // Allow painting non-uniform borders if the color and style are uniform.
-    if (_colorIsUniform && _styleIsUniform) {
-      switch (top.style) {
-        case BorderStyle.none:
-          return;
-        case BorderStyle.solid:
-          BoxBorder._paintNonUniformBorder(canvas, rect,
-              shape: shape,
-              borderRadius: borderRadius,
-              textDirection: textDirection,
-              left: left,
-              top: top,
-              right: right,
-              bottom: bottom);
-          return;
-      }
+    if (_styleIsUniform && top.style == BorderStyle.none) {
+      return;
     }
 
-    assert(() {
-      if (borderRadius != null) {
+    // Allow painting non-uniform borders if the visible colors are uniform.
+    final Set<Color> visibleColors = _distinctVisibleColors();
+    final bool hasHairlineBorder = _hasHairlineBorder;
+    // Paint a non uniform border if a single color is visible
+    // and (borderRadius is present) or (border is visible and width != 0.0).
+    if (visibleColors.length == 1 &&
+        !hasHairlineBorder &&
+        (shape == BoxShape.circle ||
+            (borderRadius != null && borderRadius != BorderRadius.zero))) {
+      BoxBorder.paintNonUniformBorder(canvas, rect,
+          shape: shape,
+          borderRadius: borderRadius,
+          textDirection: textDirection,
+          top: top.style == BorderStyle.none ? BorderSide.none : top,
+          right: right.style == BorderStyle.none ? BorderSide.none : right,
+          bottom: bottom.style == BorderStyle.none ? BorderSide.none : bottom,
+          left: left.style == BorderStyle.none ? BorderSide.none : left,
+          color: visibleColors.first);
+      return;
+    }
+
+     assert(() {
+      if (hasHairlineBorder) {
+        assert(borderRadius == null || borderRadius == BorderRadius.zero,
+            'A hairline border like `BorderSide(width: 0.0, style: BorderStyle.solid)` can only be drawn when BorderRadius is zero or null.');
```

Comment:
this new asset broke app-bar in my app after updating to 3.16.

Reply:
@Hixie @bernaferrari is it a good idea to mention this change in breaking points page for 3.16?

Reply:
Can you show the code? It isn't supposed to break anything because the behavior was wrong

Reply:
made a reproducible sample: https://dartpad.dev/?id=413770c667697ac3472fc03d12d4fe54

Issue: "FAQs" widget is not shown because of the new assert.

---

either setting the border-radius to zero or removing the BorderSide's with 0 width, fixes the problem. not sure why we were using border-side 0 in our code but code was in app-bar hence I noticed it when it glitched after stable update.

Reply:
Yeah, but the 0 as hidden is a bug, because flutter uses that for hairline border (I wish it were different but I can't change it because visual density depends on context), so instead of fixing to show it (which would be a bug for you), we fixed to assert to prevent problems in the future

Reply:
Same here. Not sure what was wrong with hairline borders, but we use them pretty extensively in our app and now they aren't visible anymore after updating to the latest Flutter version. Could you clarify what was the issue with drawing hairline borders with rounded corners?